From: Richard Acayan <mailingradian@gmail.com>
Date: Thu, 08 Dec 2022 17:43:13 -0500
Subject: [PATCH] arm64: dts: qcom: sdm670: correct cpu opps

The SoC consumes a lot more power than expected (around 0.8 W with disabled
display, mobile data, WiFi, and Bluetooth) with the original operating
performance points (OPPs) from SDM845. Add the correct OPPs for this SoC from
the downstream kernel.

Signed-off-by: Richard Acayan <mailingradian@gmail.com>
---
 arch/arm64/boot/dts/qcom/sdm670.dtsi | 284 ++++++++++++--------------------------------------
 1 file changed, 64 insertions(+), 220 deletions(-)

diff --git a/arch/arm64/boot/dts/qcom/sdm670.dtsi b/arch/arm64/boot/dts/qcom/sdm670.dtsi
index dbf469437273..23dfe4101fc1 100644
--- a/arch/arm64/boot/dts/qcom/sdm670.dtsi
+++ b/arch/arm64/boot/dts/qcom/sdm670.dtsi
@@ -13,6 +13,7 @@
 #include <dt-bindings/clock/qcom,videocc-sdm845.h>
 #include <dt-bindings/dma/qcom-gpi.h>
 #include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interconnect/qcom,osm-l3.h>
 #include <dt-bindings/interconnect/qcom,sdm670.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/phy/phy-qcom-qusb2.h>
@@ -42,7 +43,8 @@ CPU0: cpu@0 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD0>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_0>;
@@ -62,7 +64,8 @@ CPU1: cpu@100 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD1>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_100>;
@@ -79,7 +82,8 @@ CPU2: cpu@200 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD2>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_200>;
@@ -96,7 +100,8 @@ CPU3: cpu@300 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD3>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_300>;
@@ -113,7 +118,8 @@ CPU4: cpu@400 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD4>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_400>;
@@ -130,7 +136,8 @@ CPU5: cpu@500 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 0>;
 			operating-points-v2 = <&cpu0_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD5>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_500>;
@@ -147,7 +154,8 @@ CPU6: cpu@600 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 1>;
 			operating-points-v2 = <&cpu6_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD6>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_600>;
@@ -164,7 +172,8 @@ CPU7: cpu@700 {
 			enable-method = "psci";
 			qcom,freq-domain = <&cpufreq_hw 1>;
 			operating-points-v2 = <&cpu6_opp_table>;
-			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>;
+			interconnects = <&gladiator_noc MASTER_APPSS_PROC 3 &mem_noc SLAVE_EBI1 3>,
+					<&osm_l3 MASTER_OSM_L3_APPS &osm_l3 SLAVE_OSM_L3>;
 			power-domains = <&CPU_PD7>;
 			power-domain-names = "psci";
 			next-level-cache = <&L2_700>;
@@ -263,107 +272,47 @@ cpu0_opp_table: opp-table-cpu0 {
 
 		cpu0_opp1: opp-300000000 {
 			opp-hz = /bits/ 64 <300000000>;
-			opp-peak-kBps = <400000>;
+			opp-peak-kBps = <400000 4800000>;
 		};
 
-		cpu0_opp2: opp-403200000 {
-			opp-hz = /bits/ 64 <403200000>;
-			opp-peak-kBps = <400000>;
-		};
-
-		cpu0_opp3: opp-480000000 {
-			opp-hz = /bits/ 64 <480000000>;
-			opp-peak-kBps = <400000>;
-		};
-
-		cpu0_opp4: opp-576000000 {
+		cpu0_opp2: opp-576000000 {
 			opp-hz = /bits/ 64 <576000000>;
-			opp-peak-kBps = <400000>;
+			opp-peak-kBps = <400000 4800000>;
 		};
 
-		cpu0_opp5: opp-652800000 {
-			opp-hz = /bits/ 64 <652800000>;
-			opp-peak-kBps = <400000>;
-		};
-
-		cpu0_opp6: opp-748800000 {
+		cpu0_opp3: opp-748800000 {
 			opp-hz = /bits/ 64 <748800000>;
-			opp-peak-kBps = <1200000>;
-		};
-
-		cpu0_opp7: opp-825600000 {
-			opp-hz = /bits/ 64 <825600000>;
-			opp-peak-kBps = <1200000>;
-		};
-
-		cpu0_opp8: opp-902400000 {
-			opp-hz = /bits/ 64 <902400000>;
-			opp-peak-kBps = <1200000>;
-		};
-
-		cpu0_opp9: opp-979200000 {
-			opp-hz = /bits/ 64 <979200000>;
-			opp-peak-kBps = <1200000>;
+			opp-peak-kBps = <1200000 4800000>;
 		};
 
-		cpu0_opp10: opp-998400000 {
+		cpu0_opp4: opp-998400000 {
 			opp-hz = /bits/ 64 <998400000>;
-			opp-peak-kBps = <1804000>;
-		};
-
-		cpu0_opp11: opp-1056000000 {
-			opp-hz = /bits/ 64 <1056000000>;
-			opp-peak-kBps = <1804000>;
+			opp-peak-kBps = <1804000 8908800>;
 		};
 
-		cpu0_opp12: opp-1132800000 {
-			opp-hz = /bits/ 64 <1132800000>;
-			opp-peak-kBps = <1804000>;
-		};
-
-		cpu0_opp13: opp-1209600000 {
+		cpu0_opp5: opp-1209600000 {
 			opp-hz = /bits/ 64 <1209600000>;
-			opp-peak-kBps = <2188000>;
-		};
-
-		cpu0_opp14: opp-1228800000 {
-			opp-hz = /bits/ 64 <1228800000>;
-			opp-peak-kBps = <2188000>;
+			opp-peak-kBps = <2188000 8908800>;
 		};
 
-		cpu0_opp15: opp-1324800000 {
+		cpu0_opp6: opp-1324800000 {
 			opp-hz = /bits/ 64 <1324800000>;
-			opp-peak-kBps = <2188000>;
+			opp-peak-kBps = <2188000 13516800>;
 		};
 
-		cpu0_opp16: opp-1420800000 {
-			opp-hz = /bits/ 64 <1420800000>;
-			opp-peak-kBps = <2188000>;
-		};
-
-		cpu0_opp17: opp-1516800000 {
+		cpu0_opp7: opp-1516800000 {
 			opp-hz = /bits/ 64 <1516800000>;
-			opp-peak-kBps = <3072000>;
+			opp-peak-kBps = <3072000 15052800>;
 		};
 
-		cpu0_opp18: opp-1612800000 {
+		cpu0_opp8: opp-1612800000 {
 			opp-hz = /bits/ 64 <1612800000>;
-			opp-peak-kBps = <3072000>;
-		};
-
-		cpu0_opp19: opp-1689600000 {
-			opp-hz = /bits/ 64 <1689600000>;
-			opp-peak-kBps = <3072000>;
+			opp-peak-kBps = <3072000 22118400>;
 		};
 
-		cpu0_opp20: opp-1708800000 {
+		cpu0_opp9: opp-1708800000 {
 			opp-hz = /bits/ 64 <1708800000>;
-			opp-peak-kBps = <4068000>;
-		};
-
-		cpu0_opp21: opp-1766400000 {
-			opp-hz = /bits/ 64 <1766400000>;
-			opp-peak-kBps = <4068000>;
+			opp-peak-kBps = <4068000 23040000>;
 		};
 	};
 
@@ -373,167 +322,52 @@ cpu6_opp_table: opp-table-cpu6 {
 
 		cpu6_opp1: opp-300000000 {
 			opp-hz = /bits/ 64 <300000000>;
-			opp-peak-kBps = <800000>;
+			opp-peak-kBps = <400000 4800000>;
 		};
 
-		cpu6_opp2: opp-403200000 {
-			opp-hz = /bits/ 64 <403200000>;
-			opp-peak-kBps = <800000>;
-		};
-
-		cpu6_opp3: opp-480000000 {
-			opp-hz = /bits/ 64 <480000000>;
-			opp-peak-kBps = <1804000>;
-		};
-
-		cpu6_opp4: opp-576000000 {
-			opp-hz = /bits/ 64 <576000000>;
-			opp-peak-kBps = <1804000>;
-		};
-
-		cpu6_opp5: opp-652800000 {
+		cpu6_opp2: opp-652800000 {
 			opp-hz = /bits/ 64 <652800000>;
-			opp-peak-kBps = <1804000>;
-		};
-
-		cpu6_opp6: opp-748800000 {
-			opp-hz = /bits/ 64 <748800000>;
-			opp-peak-kBps = <1804000>;
+			opp-peak-kBps = <400000 4800000>;
 		};
 
-		cpu6_opp7: opp-825600000 {
+		cpu6_opp3: opp-825600000 {
 			opp-hz = /bits/ 64 <825600000>;
-			opp-peak-kBps = <2188000>;
+			opp-peak-kBps = <1200000 4800000>;
 		};
 
-		cpu6_opp8: opp-902400000 {
-			opp-hz = /bits/ 64 <902400000>;
-			opp-peak-kBps = <2188000>;
-		};
-
-		cpu6_opp9: opp-979200000 {
+		cpu6_opp4: opp-979200000 {
 			opp-hz = /bits/ 64 <979200000>;
-			opp-peak-kBps = <2188000>;
+			opp-peak-kBps = <1200000 4800000>;
 		};
 
-		cpu6_opp10: opp-1056000000 {
-			opp-hz = /bits/ 64 <1056000000>;
-			opp-peak-kBps = <3072000>;
-		};
-
-		cpu6_opp11: opp-1132800000 {
+		cpu6_opp5: opp-1132800000 {
 			opp-hz = /bits/ 64 <1132800000>;
-			opp-peak-kBps = <3072000>;
-		};
-
-		cpu6_opp12: opp-1209600000 {
-			opp-hz = /bits/ 64 <1209600000>;
-			opp-peak-kBps = <4068000>;
-		};
-
-		cpu6_opp13: opp-1286400000 {
-			opp-hz = /bits/ 64 <1286400000>;
-			opp-peak-kBps = <4068000>;
+			opp-peak-kBps = <2188000 8908800>;
 		};
 
-		cpu6_opp14: opp-1363200000 {
+		cpu6_opp6: opp-1363200000 {
 			opp-hz = /bits/ 64 <1363200000>;
-			opp-peak-kBps = <4068000>;
-		};
-
-		cpu6_opp15: opp-1459200000 {
-			opp-hz = /bits/ 64 <1459200000>;
-			opp-peak-kBps = <4068000>;
+			opp-peak-kBps = <4068000 12902400>;
 		};
 
-		cpu6_opp16: opp-1536000000 {
+		cpu6_opp7: opp-1536000000 {
 			opp-hz = /bits/ 64 <1536000000>;
-			opp-peak-kBps = <5412000>;
-		};
-
-		cpu6_opp17: opp-1612800000 {
-			opp-hz = /bits/ 64 <1612800000>;
-			opp-peak-kBps = <5412000>;
-		};
-
-		cpu6_opp18: opp-1689600000 {
-			opp-hz = /bits/ 64 <1689600000>;
-			opp-peak-kBps = <5412000>;
+			opp-peak-kBps = <4068000 12902400>;
 		};
 
-		cpu6_opp19: opp-1766400000 {
-			opp-hz = /bits/ 64 <1766400000>;
-			opp-peak-kBps = <6220000>;
+		cpu6_opp8: opp-1747200000 {
+			opp-hz = /bits/ 64 <1747200000>;
+			opp-peak-kBps = <4068000 15052800>;
 		};
 
-		cpu6_opp20: opp-1843200000 {
+		cpu6_opp9: opp-1843200000 {
 			opp-hz = /bits/ 64 <1843200000>;
-			opp-peak-kBps = <6220000>;
-		};
-
-		cpu6_opp21: opp-1920000000 {
-			opp-hz = /bits/ 64 <1920000000>;
-			opp-peak-kBps = <7216000>;
+			opp-peak-kBps = <4068000 15052800>;
 		};
 
-		cpu6_opp22: opp-1996800000 {
+		cpu6_opp10: opp-1996800000 {
 			opp-hz = /bits/ 64 <1996800000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp23: opp-2092800000 {
-			opp-hz = /bits/ 64 <2092800000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp24: opp-2169600000 {
-			opp-hz = /bits/ 64 <2169600000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp25: opp-2246400000 {
-			opp-hz = /bits/ 64 <2246400000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp26: opp-2323200000 {
-			opp-hz = /bits/ 64 <2323200000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp27: opp-2400000000 {
-			opp-hz = /bits/ 64 <2400000000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp28: opp-2476800000 {
-			opp-hz = /bits/ 64 <2476800000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp29: opp-2553600000 {
-			opp-hz = /bits/ 64 <2553600000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp30: opp-2649600000 {
-			opp-hz = /bits/ 64 <2649600000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp31: opp-2745600000 {
-			opp-hz = /bits/ 64 <2745600000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp32: opp-2803200000 {
-			opp-hz = /bits/ 64 <2803200000>;
-			opp-peak-kBps = <7216000>;
-		};
-
-		cpu6_opp33: opp-1747200000 {
-			opp-hz = /bits/ 64 <1747200000>;
-			opp-peak-kBps = <6220000>;
+			opp-peak-kBps = <6220000 19046400>;
 		};
 	};
 
@@ -2423,6 +2257,16 @@ intc: interrupt-controller@17a00000 {
 			#interrupt-cells = <3>;
 		};
 
+		osm_l3: interconnect@17d41000 {
+			compatible = "qcom,sdm845-osm-l3";
+			reg = <0 0x17d41000 0 0x1400>;
+
+			clocks = <&rpmhcc RPMH_CXO_CLK>, <&gcc GPLL0>;
+			clock-names = "xo", "alternate";
+
+			#interconnect-cells = <1>;
+		};
+
 		cpufreq_hw: cpufreq@17d43000 {
 			compatible = "qcom,cpufreq-hw";
 			reg = <0 0x17d43000 0 0x1400>, <0 0x17d45800 0 0x1400>;
